pkg/block/s3/adapter.go

```diff
@@ -321,6 +321,43 @@ func (a *Adapter) GetWalker(uri *url.URL) (block.Walker, error) {
 	return NewS3Walker(a.clients.awsSession), nil
 }
 
+// refreshClientIfNeeded ensure client has some time before its token
+// expires.  It returns the updated expiry time.  It fails with
+// ErrDoesntExpire if the client is not an Expirer,
+func refreshClientIfNeeded(ctx context.Context, client S3APIWithExpirer) (time.Time, error) {
```

Comment:
Why do we need this function?

IIUC setting `ExpiryWindow` ensures that the SDK will take care of that refresh. 
Here you basically force refresh every call, no? 

```go
credentials = SDK.AssumeRole(5 minutes window expiry) // expires in now+5 minutes 
client := FooClient(credentials...)
```

So that means that every-time you call  here it's always true, no? 
```go
expiry, err := client.ExpiresAt()
if ttl < 5*time.Minute
```
 
Reference to the other config I mean: 
```go
		WebIdentityRoleProviderOptions: func(wirp *stscreds.WebIdentityRoleProvider) {
			wirp.ExpiryWindow = renewWebIdentity
```

Reply:
Interesting and weird at the same time, thanks for clarifying that!

Reply:
I _personally_ use the word "terrifying" to describe this.